server/src/main/java/org/elasticsearch/index/mapper/SortedNumericWithOffsetsDocValuesSyntheticFieldLoader.java

```diff
@@ -0,0 +1,173 @@
+        SortedDocValues offsetDocValues = DocValues.getSorted(leafReader, offsetsFieldName);
+
+        return docValuesLoader = new NumericDocValuesWithOffsetsLoader(valueDocValues, offsetDocValues, valueWriter);
+    }
+
+    @Override
+    public boolean hasValue() {
+        return docValuesLoader != null && docValuesLoader.hasValue();
+    }
+
+    @Override
+    public void write(XContentBuilder b) throws IOException {
+        if (docValuesLoader != null) {
+            docValuesLoader.write(leafName, b);
+        }
+    }
+
+    @Override
+    public String fieldName() {
+        return fullPath;
+    }
+
+    private static final class NumericDocValuesWithOffsetsLoader implements DocValuesLoader {
+        private final SortedDocValues offsetDocValues;
+        private final SortedNumericDocValues valueDocValues;
+        private final NumericValueWriter writer;
+        private final ByteArrayStreamInput scratch = new ByteArrayStreamInput();
+
+        private boolean hasValue;
+        private boolean hasOffset;
+        private int[] offsetToOrd;
+
+        NumericDocValuesWithOffsetsLoader(
+            SortedNumericDocValues valueDocValues,
+            SortedDocValues offsetDocValues,
+            NumericValueWriter writer
+        ) {
+            this.valueDocValues = valueDocValues;
+            this.offsetDocValues = offsetDocValues;
+            this.writer = writer;
+        }
+
+        @Override
+        public boolean advanceToDoc(int docId) throws IOException {
+            hasValue = valueDocValues.advanceExact(docId);
+            hasOffset = offsetDocValues.advanceExact(docId);
+            if (hasValue || hasOffset) {
+                if (hasOffset) {
+                    int offsetOrd = offsetDocValues.ordValue();
+                    var encodedValue = offsetDocValues.lookupOrd(offsetOrd);
+                    scratch.reset(encodedValue.bytes, encodedValue.offset, encodedValue.length);
+                    offsetToOrd = FieldArrayContext.parseOffsetArray(scratch);
+                } else {
+                    offsetToOrd = null;
+                }
+                return true;
+            } else {
+                offsetToOrd = null;
+                return false;
+            }
+        }
+
+        public boolean hasValue() {
+            return hasOffset || (hasValue && valueDocValues.docValueCount() > 0);
+        }
+
+        private int count() {
+            if (offsetToOrd != null) {
+                return offsetToOrd.length;
+            } else if (hasValue) {
+                return valueDocValues.docValueCount();
+            } else {
+                return 0;
+            }
+        }
+
+        public void write(String fieldName, XContentBuilder b) throws IOException {
+            if (hasValue == false && hasOffset == false) {
+                return;
+            }
+
+            int count = count();
```

Comment:
If this class implements  `CompositeSyntheticFieldLoader.DocValuesLayer` then we don't need the logic here that determine whether something needs to be serialized as array. And we can serialize regular number values as regular number values (instead of wrapping them in an array)